numpy/ma/core.py

```diff
@@ -1613,7 +1613,11 @@ def make_mask(m, copy=False, shrink=True, dtype=MaskType):
             else:
                 result = m
         else:
-            result = np.array(m, dtype=dtype, copy=copy)
+            # existence of fields implies True
+            if m.dtype.fields and np.dtype(dtype) == np.bool_:
+                result = np.ones(m.shape, dtype=dtype)
```

Comment:
I just tried rebasing, removing this block in favor of the code from #8590, but I get a test failure with trace:
```
Traceback (most recent call last):
  File "/home/allan/nmpy_dev/build/testenv/lib/python2.7/site-packages/numpy/ma/tests/test_core.py", line 4041, in test_make_mask
    test = make_mask(mask)
  File "/home/allan/nmpy_dev/build/testenv/lib/python2.7/site-packages/numpy/ma/core.py", line 1610, in make_mask
    result = np.array(filled(m, True), copy=copy, dtype=dtype, subok=True)
ValueError: Can't cast from structure to non-structure, except if the structure only has a single field.
```

I don't have time to debug right now, but I'll get around to it sometime this week.